### PR TITLE
Next-generation CI/CD pipelines with RunsOn

### LINKED ISSUE #6306: [CI] Clean up ci_build.sh wrapper script for legibility and clarity
The `tests/ci_build/ci_build.sh` script used to be a small wrapper that ran a given command inside a Docker image. Over time, the wrapper script has become highly complex. I recently made a mistake that increased the number of times Docker images were re-built from scratch (#6305), and that one-line mistake cost us dearly, burning precious CI budget ~1.5x faster.

The wrapper script must be cleaned up for legibility and clarity.

### LINKED ISSUE #7525: Reduce the time cost for CI
Take https://github.com/dmlc/xgboost/runs/4589165334?check_suite_focus=true as an example,

CI will build the xgboost library twice for Test XGBoost4j and Test XGBoost4j-Spark.

### LINKED ISSUE #9896: Better organize scripts for CI and maintenance tasks
Currently, we update the version number manually (e.g. #9895). There should a script to automate the process.

### LINKED ISSUE #10229: [CI] Retire Mac Mini worker in BuildKite
Now that GitHub Actions provide M1 workers (`macos-14`), we don't need to maintain our own infra to test Apple Silicon.

### PR DESCRIPTION
**Pull request by hcho3**
Suggestion to reviewers: Go to https://github.com/hcho3/xgboost/tree/xgboost-ci-ng to easily browse the workflow definitions and scripts. Also pay attention to the following:
1. Overall organization of workflows
2. How containers are defined and constructed

See https://xgboost--11001.org.readthedocs.build/en/11001/contrib/ci.html for the overview.

-------

In #8142, we migrated our CI/CD pipelines from Jenkins to BuildKite. BuildKite has served us well in the past 2 years.
BuildKite was superior to Jenkins in the following ways:
* Provision of cloud components are now documented in scripts that are stored as part of the XGBoost git repository.
* We are no longer subject to frequent failures and security issues from Jenkins plug-ins.

However, over time, the BuildKite setup has entailed the following disadvantages:
1. We ended up with lots of infra scripts and glue code that provision cloud components. (See `tests/buildkite/infrastructure`.) The infra scripts require specialized knowledge of the cloud, and currently I am the only person who understands these scripts.
4. The provision of worker images was not fully automated. The setup steps were documented in bootstrap scripts, but someone had to manually run them. Also, the BuildKite's CloudFormation template required that we specify exact IDs for machine images, meaning that we had to re-provision the entire cloud stack every time we re-generate the machine images. In short: **It was cumbersome to generate new machine images**, so over time, we tended to not update machine images, which led to bit rots.
5. Developers who are familiar with GitHub Actions face substantial learning curve when configuring BuildKite, as BuildKite is a more niche application.
6. Over time, we added more and more files in `tests/buildkite` and `tests/ci_build`, and it was unclear as to what kinds of files belonged to each directory. (#9896)

I propose a new solution for hosting our CI/CD pipelines: [RunsOn](https://runs-on.com/). RunsOn is a web app that lets host GitHub Action runners on Amazon EC2. By migrating to RunsOn, we address all the issues described above.
1. We got rid of (almost) all specialized infra scripts. The only remaining infra script is the Packer build definition in `ops/packer`, but at least this bit is easy to understand for newcomers. (It basically provisions an EC2 instance, runs a few commands, and saves the machine image for later use.)
2. **It's now extremely easy to refresh worker images**: just run `packer build linux.pck.hcl`. Unlike BuildKite, RunsOn's CloudFormation template allows us to specify machine images using wildcards (not exact IDs), meaning that we can refresh machine images while keeping the rest of cloud infra the same. Furthermore, we should be able to setup an automated pipeline that runs Packer periodically.
4. **Now we can configure all CI/CD pipelines using the familiar syntax of GitHub Actions.** Developers now have access to the wealthy trove of existing workflow actions.
5. This pull request re-organizes files inside the `ops/` directory, as follows:
  * `ops/conda_env`: Definitions for Conda environments
  * `ops/packer`: Packer scripts to build machine images for Amazon EC2
  * `ops/patch`: Patch files
  * `ops/pipeline`: Shell scripts defining CI/CD pipelines. Most of these scripts can be run locally (to assist with development and debugging); those that must be run in the CI are guarded with invocation to `enforce-ci.sh`.
  * `ops/script`: Various utility scripts useful for testing.
  * `ops/docker`: Dockerfiles to define containers
  * `ops/docker_build.py` / `ops/docker_run.py`: replaces `ci_build.sh` with much more friendly UI. Allows users and CI pipelines to run arbitrary tasks inside containers.

Other improvements:
* **We now have a single YAML file (`ops/docker/ci_container.yml`) that keeps track of all build args to the containers.** 
* Windows workers now have much faster launch time, using the [Fast Launch](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/win-ami-config-fast-launch.html) feature. The launch time is reduced from 10 minutes to 2 minutes.
* RunsOn uses [Spot instances](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-spot-instances.html) automatically to achieve savings of 20-50%. [An upcoming version of RunsOn](https://github.com/runs-on/runs-on/pull/164) add the ability to retry jobs automatically should spot instances are interrupted.
* In a single workflow file, we can mix jobs that run on Microsoft-hosted and self-hosted runners, allowing us to group tasks that are related. For example, see `.github/workflows/jvm_tests.yml`.
* GitHub Actions provides nice graph visualization for pipelines
Screenshot 2024-11-20 105956
Screenshot 2024-11-20 110018

The only disadvantage of RunsOn is that we no longer have the option to require manual approvals prior to CI workflows running. So we need to retain `enforce_daily_budget.sh` for the cost control. (It's currently broken; I will fix it and bring it back in a different PR.) In addition, I temporarily disabled the Dependabot, until we figure out how to prevent the bot from filing a torrent of pull requests (which would explode our CI budget!).

Closes #10229
Closes #6306
Closes #9896
Closes #7525 (Now we build `libxgboost4j.so` once, using the cheapest instance.)

Progress towards #8311: A future pull request will automatically run `packer build` to rotate machine images.

### CODE REVIEW

**Comment by trivialfis** [comment]
Can this help https://github.com/marketplace/actions/manual-workflow-approval ?

**Comment by wbo4958** [comment]
Could you consider to create a compatible script compiling XGBoost locally in docker? Developers don't need to change any code to compile xgboost locally directly, That could be really help.

**Comment by hcho3** [comment]
> Could you consider to create a compatible script compiling XGBoost locally in docker?

It's already there (`ops/docker_run.py`). Developers should be able to use it to build XGBoost locally inside a Docker container.

> it would be great to de-couple some of the CI-specific features like uploading wheels from more general features like running tests and building wheels

Very good point. Let me extract out the CI-specific parts.

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    runs-on: ubuntu-latest
+    name: Type and format checks for the Python package
+    steps:
+      - uses: actions/checkout@v4.2.2
+        with:
+          submodules: 'true'
+      - uses: conda-incubator/setup-miniconda@v3.1.0
+        with:
+          miniforge-variant: Miniforge3
+          miniforge-version: latest
```
Comment (msarahan) [inline .github/workflows/lint.yml:61]: Since this gets repeated so much, you may want to define it as a repository environment variable: https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-deployments/managing-environments-for-deployment

Try to minimize hard-coded values where you can - especially for things that are hard to do search/replace on, like "latest"

File: .github/workflows/misc.yml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+env:
+  BRANCH_NAME: >-
+    ${{ github.event.pull_request.number && 'PR-' }}${{ github.event.pull_request.number || github.ref_name }}
+
+jobs:
+  gtest-cpu:
+    name: Test Google C++ test (CPU)
+    runs-on: macos-13
+    steps:
+      - uses: actions/checkout@v4.2.2
```
Comment (msarahan) [inline .github/workflows/misc.yml:21]: Is it important to pin to micro version? Would v4 suffice? If so, it would give you more leeway in terms of getting updates and fixes on the action.

File: .github/workflows/misc.yml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    name: Test installing XGBoost lib + building the C API demo
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        shell: bash -l {0}
+    steps:
+      - uses: actions/checkout@v4.2.2
+        with:
+          submodules: 'true'
+      - uses: conda-incubator/setup-miniconda@v3.1.0
```
Comment (msarahan) [inline .github/workflows/misc.yml:71]: This gets reused enough that I think it is worth having your own [local custom action](https://docs.github.com/en/actions/sharing-automations/creating-actions/creating-a-composite-action). It would change your code here to be something like:

```
uses: dmlc/xgboost/actions/miniforge-setup@main
```

where `actions/miniforge-setup/action.yml` would look like:
```
name: 'Setup miniforge'
inputs:
  environment_file: # this seems like the one parameter you might need to always pass in
    description: "environment file!!!"
    required: true
  <Optionally accept parameters to pass through to miniconda action, but make defaults be workable>
runs:
  using: "composite"
  steps:
   - uses: conda-incubator/setup-miniconda@v3.1.0
     with:
          miniforge-variant: Miniforge3
          miniforge-version: latest
          activate-environment: cpp_test
          environment-file: ops/conda_env/cpp_test.yml
          use-mamba: true
```

"dispatch scripts" might be helpful to decouple the version of the actions from the version of the repo. I described these in https://github.com/rapidsai/shared-actions

File: .github/workflows/misc.yml [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+          use-mamba: true
+      - name: Display Conda env
+        run: |
+          conda info
+          conda list
+      - name: Build and install XGBoost static library
+        run: |
+          mkdir build
+          cd build
+          cmake .. -DBUILD_STATIC_LIB=ON -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX -GNinja
```
[the remark this answers is not in the document]
Comment (hcho3) [reply .github/workflows/misc.yml]: I've moved all code to separate script files.

File: .github/workflows/misc.yml (hunk 2 above)
Comment (hcho3) [reply .github/workflows/misc.yml]: I revised all pins to specify the macro version only.

**Comment by hcho3** [comment]
TODOs:
- [ ] Write up a high-level overview of the CI/CD pipelines. Describe each category of files, as well as a few special primitives like docker_run.py and stash-artifacts.sh. This write-up should go under the documentation.
- [ ] Draw a Mermaid chart and post it in this GitHub issue.
- [ ] Implement custom composite actions to reduce duplicated code within workflows.

**Comment by wbo4958** [comment]
Hi @hcho3, could this issue https://github.com/dmlc/xgboost/issues/11011 be caused by your PR. please help to check it. Thx

**Comment by hcho3** [comment]
@wbo4958 Yes, I found the cause. Let me push the fix.

**Comment by hcho3** [comment]
@wbo4958 Should be fixed now

**Comment by hcho3** [comment]
Created a doc to summarize the new CI: https://xgboost--11001.org.readthedocs.build/en/11001/contrib/ci.html

File: ops/pipeline/test-jvm-gpu.sh [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+## Test JVM packages with CUDA. Note: this script assumes that
+## the user has already built libxgboost4j.so with CUDA support
+## and place it in the lib/ directory.
+
+set -euox pipefail
+
+SCALA_VERSION=2.12
```
Comment (trivialfis) [inline ops/pipeline/test-jvm-gpu.sh:8]: is it intended to hardcode the scala version here?

File: jvm-packages/xgboost4j-example/model/dump.raw.txt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+booster[0]:
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-example/model/dump.raw.txt:1]: which example is using this?

File: ops/conda_env/sdist_test.yml [hunk 7]
```diff
@@ -9,5 +9,3 @@ dependencies:
 - cmake
 - ninja
 - python-build
-- c-compiler
```
Comment (trivialfis) [inline ops/conda_env/sdist_test.yml:12]: Do we need to test installing with sdist?

File: ops/docker/extract_build_args.jq [hunk 8]
```diff
@@ -0,0 +1,8 @@
+def compute_build_args($input; $container_id):
```
Comment (trivialfis) [inline ops/docker/extract_build_args.jq:1]: Could you please share an example input and example output of this script? Nice little language, but difficult to visualize like `sed` to newbies like me.

File: ops/docker_build.sh [hunk 9]
```diff
@@ -0,0 +1,141 @@
+#!/bin/bash
```
Comment (trivialfis) [inline ops/docker_build.sh:1]: Looking at the code, it's fair to say that we don't need to interact with this script outside of CI, right?

File: ops/docker_run.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    )
+    parser.add_argument(
+        "--use-gpus",
+        action="store_true",
+        help=(
+            "Grant the container access to NVIDIA GPUs; requires the NVIDIA "
+            "Container Toolkit."
+        ),
+    )
+    parser.add_argument(
```
Comment (trivialfis) [inline ops/docker_run.py:125]: Thank you for having this option!

File: ops/packer/linux/bootstrap.sh [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
+  && curl -s -L https://nvidia.github.io/libnvidia-container/stable/deb/nvidia-container-toolkit.list | \
+    sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
+    sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
+sudo apt-get update
+sudo apt-get install -y nvidia-container-toolkit
+sudo nvidia-ctk runtime configure --runtime=docker
+sudo systemctl restart docker
+
+sleep 10
```
Comment (trivialfis) [inline ops/packer/linux/bootstrap.sh:35]: what's this for?

File: ops/packer/windows/windows.pkr.hcl [hunk 12]
```diff
@@ -0,0 +1,90 @@
+packer {
```
Comment (trivialfis) [inline ops/packer/windows/windows.pkr.hcl:1]: any recommended reading for this? (along with the linux version).

File: ops/docker_build.sh (hunk 9 above)
Comment (hcho3) [reply ops/docker_build.sh:1]: It's still useful as a shortcut for `docker_build.py`.

File: ops/packer/linux/bootstrap.sh (hunk 11 above)
Comment (hcho3) [reply ops/packer/linux/bootstrap.sh:35]: Just in case if the Docker daemon doesn't fully come up

File: ops/conda_env/sdist_test.yml (hunk 7 above)
Comment (hcho3) [reply ops/conda_env/sdist_test.yml:12]: Yes, in the case, I messed up the packaging and broke the sdist.

File: ops/packer/windows/windows.pkr.hcl (hunk 12 above)
Comment (hcho3) [reply ops/packer/windows/windows.pkr.hcl:1]: Sure, start with

* https://runs-on.com/guides/building-custom-ami-with-packer/
* https://developer.hashicorp.com/packer/tutorials/aws-get-started/aws-get-started-build-image

File: ops/docker/extract_build_args.jq (hunk 8 above)
Comment (hcho3) [reply ops/docker/extract_build_args.jq:1]: I will add it as part of the docstring.

File: ops/pipeline/test-jvm-gpu.sh (hunk 5 above)
Comment (hcho3) [reply ops/pipeline/test-jvm-gpu.sh:8]: Yes, I chose to only test for 2.12 here.
@wbo4958 Should we test xgboost4j-spark-gpu for both Scala 2.12 and 2.13 ?

Comment (wbo4958) [reply ops/pipeline/test-jvm-gpu.sh:8]: Yeah, xgboost4j-spark-gpu should support both of 2.12 and 2.13

**Comment by wbo4958** [comment]
Hi @hcho3,

https://github.com/dmlc/xgboost/issues/11015 may be caused by this PR

**Comment by hcho3** [comment]
@wbo4958 My bad, let me fix it.

**Comment by trivialfis** [comment]
Does this PR affect the upcoming patch release?

**Comment by hcho3** [comment]
> Does this PR affect the upcoming patch release?

No, it does not.

File: ops/pipeline/test-jvm-gpu.sh (hunk 5 above)
Comment (hcho3) [reply ops/pipeline/test-jvm-gpu.sh:8]: Got it, I will add test for both 2.12 and 2.13.

**Comment by hcho3** [comment]
https://github.com/dmlc/xgboost/issues/10933 might be due to the particular setup of the RunsOn stack. Today I am seeing random occurrence of error `cudaErrorNoDevice no CUDA-capable device is detected` in the JVM tests: https://github.com/dmlc/xgboost/actions/runs/12101273094/job/33741364702. For now I am adding `--privileged` flag to Docker runs and see if that makes the error go away.

File: .github/workflows/python_wheels_macos.yml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    name: Build wheel for ${{ matrix.platform_id }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+        - os: macos-13
+          platform_id: macosx_x86_64
+        - os: macos-14
+          platform_id: macosx_arm64
```
Comment (jameslamb) [inline .github/workflows/python_wheels_macos.yml:31]: Commenting here as one example of something I'm confused about.... am I right that the jobs in this `python_wheels_macos.yml` workflow are running on GitHub-hosted runners, not anything managed by RunsOn?

If so, how is this new file related to this PR's goals?

I'm confused enough by this that I'm worried I'm misunderstanding something foundational about this PR 😬

**Review by jameslamb — COMMENTED** [review]
I tried to review this today by reading the following:

* the PR description
* the prior discussions on the PR
* the docs at https://xgboost--11001.org.readthedocs.build/en/11001/contrib/ci.html (thanks for that link!)
* the diff

But I'm sorry, this diff is just too large for me to understand what's going on. Is it possible to make this smaller, for example to only modify the workflows that affect the Python package? And to move to a separate PR seemingly unrelated changes like whitespace and adding type hints to Python code?

Things I haven't been able to figure out in this first pass:

* where / when are docker images and VM images built?
  - *every commit to `master`? on a schedule? manually by maintainers?*
* what's the relationship between the VM images (built with `packer`) and the container images (built with `docker` and pushed to ECR)?
* which logs from this PR should I be looking at to see jobs run on RunsOn?
* I still see BuildKite check statuses on this PR... is that expected?

image

Sorry for not leaving a more helpful review. If there are smaller, more self-contained pieces of this you want my feedback on please point me to them and I'd be happy to review.

File: .github/workflows/python_wheels_macos.yml (hunk 13 above)
Comment (hcho3) [reply .github/workflows/python_wheels_macos.yml:31]: We have a mix of jobs, some using self-hosted runners and others using Microsoft-hosted runners. I will update the doc to explain how self-hosted runners are indicated in the workflow YAML syntax.

**Comment by hcho3** [comment]
> where / when are docker images and VM images built?

VM images are built manually, by running `packer build linux.pkr.hcl`.
On the other hand, Docker images are built much more frequently, at every commit and with every pull request. 

> what's the relationship between the VM images (built with packer) and the container images (built with docker and pushed to ECR)?

VM images are expected to be refreshed sparingly, whereas container images are to be refreshed at every commit. The VM image contains the minimal set of drivers and system software so that it can run the containers (possibly with access to GPUs).

> which logs from this PR should I be looking at to see jobs run on RunsOn?

Jobs that run on RunsOn will have the following entry:
Screenshot 2024-12-02 192555

> I still see BuildKite check statuses on this PR... is that expected?

Yes, BuiltKite status checks will remain until we uninstall BuildKite completely from the repo.

**Review by trivialfis — APPROVED** [review]

File: .github/workflows/python_wheels_macos.yml (hunk 13 above)
Comment (jameslamb) [reply .github/workflows/python_wheels_macos.yml:31]: I understand that there will be a mix.

But how is this new file related to this PR's goals? It seems to only contain jobs that run on Microsoft-hosted runners, and looks almost identical to what already exists on `master`:

https://github.com/dmlc/xgboost/blob/bb2e701bf30fe75a0a4b48e599a7d64020325727/.github/workflows/python_wheels.yml#L15-L26

Comment (hcho3) [reply .github/workflows/python_wheels_macos.yml:31]: Not sure why the diff is larger. Essentially `python_wheels_macos.yml` is a renamed version of `python_wheels.yml`. I rename the workflow to make it clearer about what it does.

Comment (jameslamb) [reply .github/workflows/python_wheels_macos.yml:31]: Ok, I see. I probably would have left such changes out of this PR and #11001, to help reviewers focus on the core RunsOn-related changes.

**Comment by hcho3** [comment]
Merging this early so that we can start monitoring for potential issues. We will continue our design discussion in #11046.
